src/fides/api/util/fuzzy_search_utils.py

```diff
@@ -0,0 +1,92 @@
+    """Returns whether the decrypted identity cache has expired"""
+    cache: FidesopsRedis = get_cache()
+    result = cache.get(DECRYPTED_IDENTITY_CACHE_KEY)
+    if result:
+        return False
+    return True
+
+
+def _add_decrypted_identities_to_automaton(
+    identities: Optional[Dict[str, Any]],
+    request_id: str,
+    automaton: ahocorasick.Automaton,  # pylint: disable=c-extension-no-member
+) -> None:
+    if not identities or not identities.items():
+        return
+    for key, value in identities.items():  # pylint: disable=W0612
+        if value:
+            if automaton.exists(value):
+                existing: List[str] = automaton.get(value)
+                existing.append(str(request_id))
+                # overwrites the previously found key, updates with new request id
+                automaton.add_word(value, existing)
+            else:
+                automaton.add_word(value, [request_id])
+
+
+def build_decrypted_identities_automaton(
+    query: Query,
+) -> ahocorasick.Automaton:  # pylint: disable=c-extension-no-member
+    """
+    Retrieve identities from cache. If cache is expired, retrieves from DB and caches results.
+    # Stores in automaton with format: {"decrypted identity val", ["req_id_1", "req_id_2"]}
+    """
+    automaton = ahocorasick.Automaton()  # pylint: disable=c-extension-no-member
+
+    all_privacy_requests: List[PrivacyRequest] = query.all()
+    if get_has_identity_cache_expired():
+        logger.info(
+            "Decrypted identity cache does not exist or has expired. Proceeding to cache all decrypted identities..."
+        )
+        for request in all_privacy_requests:
+            _add_decrypted_identities_to_automaton(request.get_persisted_identity().__dict__, request.id, automaton)  # type: ignore
+            # Write to cache for later
+            request.cache_decrypted_identities_by_privacy_request()
+        set_decrypted_identity_cache_signal()
+
+    else:
+        # Get identities from cache
+        for request in all_privacy_requests:
+            all_identities_for_request: Optional[Dict[str, Any]] = (
+                request.retrieve_decrypted_identities_from_cache_by_privacy_request()
+            )
+            if not all_identities_for_request:
+                # Safeguard if specific privacy requests do not have identities stored in the cache
+                logger.info(
+                    "Decrypted identities could not be found in cache for privacy request. Proceeding to look up in DB instead of using cache."
+                )
+                _add_decrypted_identities_to_automaton(request.get_persisted_identity.__dict__, request.id, automaton)  # type: ignore
+                # Write to cache for later
+                request.cache_decrypted_identities_by_privacy_request()
+            _add_decrypted_identities_to_automaton(all_identities_for_request, request.id, automaton)  # type: ignore
+    return automaton
```

Comment:
I was thinking you would follow a singleton approach like we do for our Redis cache https://github.com/ethyca/fides/blob/main/src/fides/api/util/cache.py#L124, it would be a huge speed-up if we can build this up once on system startup and just refresh it periodically or add to it as new requests come in

Reply:
Ah so you mean build the Automaton only once as well. Lemme think about how best to do this...